Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm

```diff
@@ -542,6 +551,119 @@ static inline String toMIMEType(WebExtensionTab::ImageFormat format)
     }
 }
 
+void WebExtensionContext::tabsExecuteScript(WebPageProxyIdentifier webPageProxyIdentifier, std::optional<WebExtensionTabIdentifier> tabIdentifier, const WebExtensionScriptInjectionParameters& parameters, CompletionHandler<void(std::optional<InjectionResults>, WebExtensionTab::Error)>&& completionHandler)
+{
+    auto tab = getTab(webPageProxyIdentifier, tabIdentifier);
+    if (!tab) {
+        completionHandler(std::nullopt, toErrorString(@"tabs.executeScript()", nil, @"tab not found"));
+        return;
+    }
+
+    auto *webView = tab->mainWebView();
+    if (!webView) {
+        completionHandler(std::nullopt, toErrorString(@"tabs.executeScript()", nil, @"could not execute script in tab"));
+        return;
+    }
+
+    if (!hasPermission(webView.URL, tab.get())) {
+        completionHandler(std::nullopt, toErrorString(@"tabs.executeScript()", nil, @"this extension does not have access to this tab"));
+        return;
+    }
+
+    std::optional<SourcePair> scriptData;
+    if (parameters.code)
+        scriptData = make_pair(parameters.code.value(), std::nullopt);
+    else {
+        NSString *filePath = parameters.files.value().first();
+        scriptData = sourcePairForResource(filePath, m_extension);
+        if (!scriptData) {
+            completionHandler(std::nullopt, toErrorString(@"tabs.executeScript()", nil, @"Invalid resource: %@", filePath));
+            return;
+        }
+    }
+
+    auto injectionResults = InjectionResultHolder::create();
+    auto aggregator = MainRunLoopCallbackAggregator::create([injectionResults, completionHandler = WTFMove(completionHandler)]() mutable {
+        completionHandler(injectionResults->results, std::nullopt);
+    });
+
+    [webView _frames:makeBlockPtr([this, protectedThis = Ref { *this }, webView = RetainPtr { webView }, tab = RefPtr { tab }, injectionResults, aggregator = WTFMove(aggregator), parameters = WTFMove(parameters)](_WKFrameTreeNode *mainFrame) mutable {
```

Comment:
Should be able to do `tab` instead of `tab = RefPtr { tab }` and `aggregator` instead of `aggregator = WTFMove(aggregator)`. Ditto for `parameters` instead of `parameters = WTFMove(parameters)`.